KAFKA-15443: Upgrade RocksDB to 9.7.3

swikarpat opened the pull request:
Upgrades RocksDB to version 9.7.3 to address Jira [KAFKA-15443](https://issues.apache.org/jira/browse/KAFKA-15443)

This PR addresses the following compatibility issues introduced by the RocksDB upgrade:

*   **Removal of `AccessHint`:** The `AccessHint` class was completely removed in RocksDB 9.7.3. This required removing all import statements, variable declarations, method parameters, method return types, and static method calls related to `AccessHint` in `RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java `RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.java Unused methods are removed in RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java

*   **Removal of `NO_FILE_CLOSES`:** The `NO_FILE_CLOSES` metric was also removed in RocksDB 9.7.3. The calculation for `numberOfOpenFiles` in `RocksDBMetricsRecorder.java` has been adjusted to now track the total number of file opens since the last reset. The previous calculation, which subtracted `NO_FILE_CLOSES` from `NO_FILE_OPENS`, is no longer possible.
The reason RocksDB team removed NO_FILE_CLOSES it seems to me they detected not properly working: https://github.com/search?q=repo%3Afacebook%2Frocksdb+NO_FILE_CLOSES&type=issues

Any feedback or any proposed changes you would like me to do are welcome let me know.
@cadonna @mjsax 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

swikarpat committed:
KAFKA-15443: Upgrade RocksDB to 9.7.3

This PR addresses the following compatibility issues introduced by the RocksDB upgrade:

Removal of AccessHint: The AccessHint class was completely removed in RocksDB 9.7.3. This required removing all import statements, variable declarations, method parameters, method return types, and static method calls related to AccessHint in RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.java Unused methods are removed in RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java

Removal of NO_FILE_CLOSES: The NO_FILE_CLOSES metric was also removed in RocksDB 9.7.3. The calculation for numberOfOpenFiles in RocksDBMetricsRecorder.java has been adjusted to now track the total number of file opens since the last reset. The previous calculation, which subtracted NO_FILE_CLOSES from NO_FILE_OPENS, is no longer possible.
The reason RocksDB team removed NO_FILE_CLOSES it seems to me they detected not properly working: https://github.com/search?q=repo%3Afacebook%2Frocksdb+NO_FILE_CLOSES&type=issues

KAFKA-15443: Upgrade RocksDB to 9.7.3

Update docs/ops.html

Co-authored-by: Bruno Cadonna <cadonna@apache.org>

Update docs/streams/upgrade-guide.html

Co-authored-by: Bruno Cadonna <cadonna@apache.org>

Update docs/streams/upgrade-guide.html

Co-authored-by: Bruno Cadonna <cadonna@apache.org>

Update upgrade-guide.html

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/metrics/RocksDBMetricsRecorder.java:
```diff
@@ -462,8 +462,10 @@ public void record(final long now) {
+
+            // NO_FILE_CLOSES metric was removed in RocksDB 9.7.3.
+            // This now tracks the total number of file opens since the last reset.
+            numberOfOpenFiles += valueProviders.statistics.getAndResetTickerCount(TickerType.NO_FILE_OPENS);
```
As commented on the Jira ticket, changing semantics of an existing metric does not sound right. I would rather just report `-1` surrogate for now, and deprecate this metric with 4.0 release.

Curious to hear what others think.

ableegoldman replied:
Tend to agree that changing semantics is no good, but it sounds like the new metric will be useful -- is it possible to deprecate the old one and introduce a new one that corresponds to the same underlying thing? I really don't know anything about how we expose the RocksDB metrics though so Bruno is probably better suited to chime in

swikarpat replied:
@mjsax Thanks for the suggestion. I'm not entirely sure how to implement reporting a -1 surrogate and deprecating the metric for the 4.0 release. Could you provide some guidance or point me to relevant examples? NO_FILE_CLOSES metric declared in TickerType.class inside org.rocksdb package. We can't modify that class, can we? @cadonna @ableegoldman

cadonna replied:
@swikarpat I believe @mjsax means to not return the number of opened files but just `-1`. I agree with him. The number of opened files without the number of closed files does not give us much info. Furthermore, by recording the number of opened files, the metric would seem correct, but actually would not show the correct values because the metric is defined as the number of currently open files. Returning constant `-1` would make it clear that the metric does not work.
In addition to returning `-1`, could you please also update the documentation of the metrics in this PR: https://github.com/apache/kafka/blob/a0a501952b6d61f6f273bdb8f842346b51e9dfce/docs/ops.html#L3457

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java:
```diff
@@ -842,12 +830,6 @@ public Options setSstFileManager(final SstFileManager sstFileManager) {
 
-    @Override
-    public Options setLogger(final org.rocksdb.Logger logger) {
-        dbOptions.setLogger(logger);
```
If this was removed, too, can you add it to the PR description.

swikarpat replied:
@mjsax thanks for feedback. I wanted to confirm setLogger() method. Parameter inside setLogger() has changed to LoggerInterface. I've made corrections and put method back in that java class.

cadonna replied:
It is fine to change it like you did. If users use a logger class that is derived from `Logger` they have to modify it in any case because `Logger` now implements interface `LoggerInterface`. Thus, the change of the method will most likely not directly affect users on a source level. It will affect them on binary level, though, but that is the trade-off here.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java:
```diff
@@ -571,17 +570,6 @@ public long dbWriteBufferSize() {
-
-    @Override
-    public AccessHint accessHintOnCompactionStart() {
-        return dbOptions.accessHintOnCompactionStart();
```
It's too bad that RockDB does not mark methods as deprecated but just removes stuff.

This was always an issue for us (even if we have to admit, that we are upgrade from a quite old version...). I guess we just need to bite the bullet and make this breaking change... Seem we need to start tracking RocksDB changes more closely and frequently and do KIPs early on to prepare for breaking changes. (What basically implies, that we can upgrade RocksDB only in major release if there are breaking changes...)

Or we change our "policy" for RocksDB compatibility and document it properly, reserving the right to break compatibility via RocksDB version bumps also in minor release.

ableegoldman replied:
Yeah this drives me crazy. Personally I think it's ok to bite the bullet on this one but let's definitely come up with a strategy for this going forward. I'd say a good compromise is to meet in the middle and do both, that is, publish an official policy for RocksDB APIs which says there is no guarantee of the usual one-year deprecation & only major version removal, and any deprecations are best-effort. I'm less worried about the deprecations and more about restricting our ability to upgrade rocksdb to only major releases.

At the same time, I think we can and should do better. I like the idea of  looking ahead at each new RocksDB release to mark any changed or removed APIs exposed to our users so we can deprecate them ASAP. We can hash out the details of how often and how to share this responsibility later (for example maybe this should be one of the Streams RM responsibilities, to look for upcoming API changes in future rocks and make sure they get deprecated in that release

mjsax reviewed (commented):
Thanks for the PR. I think we need to bite the bullet and merge this as-is. Given that we are heading into a major release, it should hopefully be ok.

Curious to hear how others think about it. \cc @cadonna @ableegoldman

cadonna reviewed (commented):
Thanks for the PR @swikarpat !

Besides methods that were removed, there is at least one method that was added, i.e., `setMemtableMaxRangeDeletions()`. That is also the reason why `RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest` fails.

In the compatibility report you attached to the ticket there are quite some compatibility issues listed. IMO, we should state the most important ones on the upgrade notes and link to the compatibility report for details. Could you please update the [upgrade notes ](https://github.com/apache/kafka/blob/d208abb38eac87e6cb534bb5eadbf4aedd3a9e06/docs/streams/upgrade-guide.html#L37) with this info in this PR?

I left some more comments on the PR.

swikarpat commented:
@cadonna I have added methods and also made changes in Test class and also documents. Unit Tests keep failing. Though I ran unit tests for given two Test file and ran passed. Can you take a look what went wrong and feedback? Thanks @mjsax

mjsax commented:
RocksDB did push out newer release since we started this PR. There is now 9.10.0 -- not sure if we want to go with a .0 version, but there is also 9.9.3. -- Should we use the latest version instead of 9.7.3?

mjsax commented:
> Unit Tests keep failing. Though I ran unit tests for given two Test file and ran passed. Can you please take a look what went wrong and feedback?

No need to worry about these. We know they are flaky, and AFAIK somebody is already working on a fix. As long as no test related to this PR fail, we can merge anyway. Btw: in case a test get's fixed, you would need to rebase your branch to `trunk` to pickup the fix.

mjsax commented on docs/streams/upgrade-guide.html:
```diff
@@ -139,8 +139,6 @@ <h3 class="anchor-heading"><a id="streams_notable_changes" class="anchor-link"><
         More details about the new config <code>StreamsConfig#TOPOLOGY_OPTIMIZATION_CONFIG</code> can be found in <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-295%3A+Add+Streams+Configuration+Allowing+for+Optional+Topology+Optimization">KIP-295</a>.
     </p>
 
-    <h3><a id="streams_api_changes_400" href="#streams_api_changes_400">Streams API changes in 4.0.0</a></h3>
```
Why did you move this line?

cadonna replied:
Still not clear why this line was moved.

mjsax commented on docs/streams/upgrade-guide.html:
```diff
@@ -173,6 +171,22 @@ <h3><a id="streams_api_changes_400" href="#streams_api_changes_400">Streams API
+    <h3><a id="streams_api_changes_400" href="#streams_api_changes_390">Streams API changes in 4.0.0</a></h3>
+
+    <p>
+        * Upgraded RocksDB dependency to version 9.7.3 (from 7.9.2). This upgrade incorporates various improvements and optimizations within RocksDB. However, it also introduces some API changes.
```
Why are there `*` at the beginning of each line? This is regular HTML inside a <p> block, so having `*` seem not to be right?

If you want to use bullet points, you need to use the corresponding HTML markup.

mjsax commented on docs/streams/upgrade-guide.html:
```diff
@@ -173,6 +171,22 @@ <h3><a id="streams_api_changes_400" href="#streams_api_changes_400">Streams API
+    <p>
+        * Upgraded RocksDB dependency to version 9.7.3 (from 7.9.2). This upgrade incorporates various improvements and optimizations within RocksDB. However, it also introduces some API changes.
+        * API Changes in the RocksDB class:
+        * Removed Class: The org.rocksdb.AccessHint class, along with its associated methods, has been removed.
```
```suggestion
        * Removed Class: The <code>org.rocksdb.AccessHint</code> class, along with its associated methods, has been removed.
```

mjsax replied:
Similar elsewhere

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java:
```diff
@@ -1135,6 +1136,7 @@ public Options setBloomLocality(final int bloomLocality) {
         return this;
     }
 
+
```
nit: avoid unnecessary whitespace changes

mjsax commented on streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.java:
```diff
@@ -126,6 +127,7 @@ public void shouldOverwriteAllOptionsMethods() throws Exception {
         }
     }
 
+
```
nit. as above

swikarpat commented:
> RocksDB did push out newer release since we started this PR. There is now 9.10.0 -- not sure if we want to go with a .0 version, but there is also 9.9.3. -- Should we use the latest version instead of 9.7.3?



> RocksDB did push out newer release since we started this PR. There is now 9.10.0 -- not sure if we want to go with a .0 version, but there is also 9.9.3. -- Should we use the latest version instead of 9.7.3?
Thanks for feedback. Yes I am aware of that but only unto 9.7.3 currently published on maven central. I've asked RocksDB team for timeline to publish latest version on maven central as here https://github.com/facebook/rocksdb/issues/13214

mjsax reviewed (approved):
LGTM.

swikarpat commented:
> > Unit Tests keep failing. Though I ran unit tests for given two Test file and ran passed. Can you please take a look what went wrong and feedback?
> 
> No need to worry about these. We know they are flaky, and AFAIK somebody is already working on a fix. As long as no test related to this PR fail, we can merge anyway. Btw: in case a test get's fixed, you would need to rebase your branch to `trunk` to pickup the fix.

@mjsax I've rebased onto the latest trunk and squashed my commits into one.

swikarpat commented:
> LGTM.

@mjsax Now I have rebased onto the latest `trunk` and squashed my commits into one.

swikarpat commented:
@mjsax Now I have rebased onto the latest trunk and squashed my commits into one. Could you merge this PR as I do not have access to merge. Thanks

cadonna commented on docs/streams/upgrade-guide.html:
```diff
@@ -173,6 +171,21 @@ <h3><a id="streams_api_changes_400" href="#streams_api_changes_400">Streams API
+        <code>org.rocksdb.Options.memtableMaxRangeDeletions(): int</code> has been added, allowing retrieval of the configured maximum number of range deletions in the memtable.
+
+        The <code>org.rocksdb.Options.setLogger()</code> method now accepts a <code>LoggerInterface</code> as a parameter instead of the previous <code>Logger</code>.
+        The <code>NO_FILE_CLOSES</code> field has been removed from the <code>org.rocksdb.TickerTypeenum</code>.
```
Could you please mention that since this field is removed, the `number-open-files` metrics does not work as expected?

cadonna commented on docs/ops.html:
```diff
@@ -3473,11 +3473,6 @@ <h5 class="anchor-heading"><a id="kafka_streams_rocksdb_monitoring" class="ancho
       <td>The maximum duration in ms of disc compactions.</td>
       <td>kafka.streams:type=stream-state-metrics,thread-id=([-.\w]+),task-id=([-.\w]+),[store-scope]-id=([-.\w]+)</td>
     </tr>
-    <tr>
-      <td>number-open-files</td>
-      <td>The number of current open files.</td>
-      <td>kafka.streams:type=stream-state-metrics,thread-id=([-.\w]+),task-id=([-.\w]+),[store-scope]-id=([-.\w]+)</td>
-    </tr>
```
Instead of removing this row, could you add a comment that the metric will return constant `-1` and why it returns that?
In the end the metric is still exposed, it just returns a useless constant value.

If you want, you are welcome to create a KIP to remove this metric from Streams. Once that KIP is accepted, we can remove the metric from in the next major release.

cadonna commented on docs/streams/upgrade-guide.html:
```diff
@@ -173,6 +171,21 @@ <h3><a id="streams_api_changes_400" href="#streams_api_changes_400">Streams API
+        Upgraded RocksDB dependency to version 9.7.3 (from 7.9.2). This upgrade incorporates various improvements and optimizations within RocksDB. However, it also introduces some API changes.
+        The <code>org.rocksdb.AccessHint</code> class, along with its associated methods, has been removed.
+
+        <code>org.rocksdb.Options.setMemtableMaxRangeDeletions(int p1): ColumnFamilyOptionsInterface</code> has been added, providing control over the maximum number of range deletions in the memtable.
+        <code>org.rocksdb.Options.memtableMaxRangeDeletions(): int</code> has been added, allowing retrieval of the configured maximum number of range deletions in the memtable.
```
I would only add statements about stuff that was removed because that breaks the users' applications.

cadonna commented on docs/streams/upgrade-guide.html:
```diff
@@ -173,6 +171,21 @@ <h3><a id="streams_api_changes_400" href="#streams_api_changes_400">Streams API
+        The <code>org.rocksdb.Options.setLogger()</code> method now accepts a <code>LoggerInterface</code> as a parameter instead of the previous <code>Logger</code>.
+        The <code>NO_FILE_CLOSES</code> field has been removed from the <code>org.rocksdb.TickerTypeenum</code>.
+        Some data types used in RocksDB's Java API have been modified. These changes, along with the removed class, field, and new methods, are primarily relevant to users implementing custom RocksDB configurations.
+        These changes are expected to be largely transparent to most Kafka Streams users. However, those employing advanced RocksDB customizations within their Streams applications, particularly through the <code>rocksdb.config.setter</code>, are advised to consult the detailed RocksDB 9.7.3 changelog to ensure a smooth transition and adapt their configurations as needed. Specifically, users leve ...
```
The compatibility report also lists the removal of methods of `BlockBasedTableConfig` which might be important here.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java:
There are a couple of deprecated methods in this file that can be completely removed since 4.0 is a major release. For example `baseBackgroundCompactions()`.

cadonna reviewed (commented):
Thanks for the updates, @swikarpat !

I have a couple of comments.

swikarpat commented:
@cadonna Thanks for your feedback. I've made changes accordingly. Let me know if anything .. Thanks.

cadonna commented on docs/ops.html:
```diff
@@ -3475,7 +3475,7 @@ <h5 class="anchor-heading"><a id="kafka_streams_rocksdb_monitoring" class="ancho
     <tr>
       <td>number-open-files</td>
-      <td>The number of current open files.</td>
+      <td>This metric will return constant -1 and no longer reliable because it depended on NO_FILE_CLOSES, which was removed in RocksDB 9.7.3</td>
```
```suggestion
      <td>This metric will return constant -1 because the RocksDB's counter NO_FILE_CLOSES has been removed in RocksDB 9.7.3</td>
```

cadonna commented on docs/streams/upgrade-guide.html:
```diff
@@ -173,6 +173,16 @@ <h3><a id="streams_api_changes_400" href="#streams_api_changes_400">Streams API
+    <p>
+        Upgraded RocksDB dependency to version 9.7.3 (from 7.9.2). This upgrade incorporates various improvements and optimizations within RocksDB. However, it also introduces some API changes.
+        The <code>org.rocksdb.AccessHint</code> class, along with its associated methods, has been removed.
+        The <code>org.rocksdb.BlockBasedTableConfig</code> class, along with its associated methods, has been removed.
```
`BlockBasedTableConfig` has not been removed. Just a couple of methods on this class have been removed.

cadonna commented on docs/streams/upgrade-guide.html:
```diff
@@ -173,6 +173,16 @@ <h3><a id="streams_api_changes_400" href="#streams_api_changes_400">Streams API
+        Upgraded RocksDB dependency to version 9.7.3 (from 7.9.2). This upgrade incorporates various improvements and optimizations within RocksDB. However, it also introduces some API changes.
+        The <code>org.rocksdb.AccessHint</code> class, along with its associated methods, has been removed.
+        The <code>org.rocksdb.BlockBasedTableConfig</code> class, along with its associated methods, has been removed.
+        The <code>NO_FILE_CLOSES</code> field has been removed from the <code>org.rocksdb.TickerTypeenum</code> as a result the <code>number-open-files</code> metrics does not work as expeced.
```
```suggestion
        The <code>NO_FILE_CLOSES</code> field has been removed from the <code>org.rocksdb.TickerTypeenum</code> as a result the <code>number-open-files</code> metrics does not work as expected. Metric <code>number-open-files</code> returns constant -1 from now on until it will officially be removed.
```

cadonna commented on docs/streams/upgrade-guide.html:
```diff
@@ -173,6 +173,16 @@ <h3><a id="streams_api_changes_400" href="#streams_api_changes_400">Streams API
+        The <code>NO_FILE_CLOSES</code> field has been removed from the <code>org.rocksdb.TickerTypeenum</code> as a result the <code>number-open-files</code> metrics does not work as expeced.
+        The <code>org.rocksdb.Options.setLogger()</code> method now accepts a <code>LoggerInterface</code> as a parameter instead of the previous <code>Logger</code>.
+        Some data types used in RocksDB's Java API have been modified. These changes, along with the removed class, field, and new methods, are primarily relevant to users implementing custom RocksDB configurations.
+        These changes are expected to be largely transparent to most Kafka Streams users. However, those employing advanced RocksDB customizations within their Streams applications, particularly through the <code>rocksdb.config.setter</code>, are advised to consult the detailed RocksDB 9.7.3 changelog to ensure a smooth transition and adapt their configurations as needed. Specifically, users leve ...
```
```suggestion
        These changes are expected to be largely transparent to most Kafka Streams users. However, those employing advanced RocksDB customizations within their Streams applications, particularly through the <code>rocksdb.config.setter</code>, are advised to consult the detailed RocksDB 9.7.3 changelog to ensure a smooth transition and adapt their configurations as needed. Specifically, users leveraging the removed <code>AccessHint</code> class, the removed methods from the <code>BlockBasedTableConfig</code> class, the <code>NO_FILE_CLOSES</code> field from <code>TickerType</code>, or relying on the previous signature of <code>setLogger()</code> will need to update their implementations.
```

cadonna reviewed (commented):
Thanks for the updates, @swikarpat !

I have a couple of minor comments!

swikarpat commented:
@cadonna Thanks for feedback. I've adjusted your suggested changes accordingly.

cadonna reviewed (approved):
Thanks for the updates, @swikarpat !

LGTM!

I restarted the failed job. Once that terminates, I am going to merge this PR.

swikarpat commented:
﻿Hello Matthias,Rocksdb 9.10.0 version now available they published yesterday on Maven central.Do we want to upgrade to 9.10.0 from 9.7.3 (current)ThanksSwikarOn Jan 6, 2025, at 3:00 AM, Bruno Cadonna ***@***.***> wrote:﻿
@cadonna commented on this pull request.

Thanks for the updates, @swikarpat !
I have a couple of comments.

In docs/streams/upgrade-guide.html:
> @@ -173,6 +171,21 @@ <h3><a id="streams_api_changes_400" href="#streams_api_changes_400">Streams API
       See <a href="https://cwiki.apache.org/confluence/x/TZCMEw">KIP-1112</a> for more details.
     </p>
 
+    <h3><a id="streams_api_changes_400" href="#streams_api_changes_400">Streams API changes in 4.0.0</a></h3>
+
+    <p>
+        Upgraded RocksDB dependency to version 9.7.3 (from 7.9.2). This upgrade incorporates various improvements and optimizations within RocksDB. However, it also introduces some API changes.
+        The <code>org.rocksdb.AccessHint</code> class, along with its associated methods, has been removed.
+
+        <code>org.rocksdb.Options.setMemtableMaxRangeDeletions(int p1): ColumnFamilyOptionsInterface</code> has been added, providing control over the maximum number of range deletions in the memtable.
+        <code>org.rocksdb.Options.memtableMaxRangeDeletions(): int</code> has been added, allowing retrieval of the configured maximum number of range deletions in the memtable.
+
+        The <code>org.rocksdb.Options.setLogger()</code> method now accepts a <code>LoggerInterface</code> as a parameter instead of the previous <code>Logger</code>.
+        The <code>NO_FILE_CLOSES</code> field has been removed from the <code>org.rocksdb.TickerTypeenum</code>.

Could you please mention that since this field is removed, the number-open-files metrics does not work as expected?



In docs/ops.html:
> -    <tr>
-      <td>number-open-files</td>
-      <td>The number of current open files.</td>
-      <td>kafka.streams:type=stream-state-metrics,thread-id=([-.\w]+),task-id=([-.\w]+),[store-scope]-id=([-.\w]+)</td>
-    </tr>

Instead of removing this row, could you add a comment that the metric will return constant -1 and why it returns that?
In the end the metric is still exposed, it just returns a useless constant value.
If you want, you are welcome to create a KIP to remove this metric from Streams. Once that KIP is accepted, we can remove the metric from in the next major release.



In docs/streams/upgrade-guide.html:
> +        <code>org.rocksdb.Options.setMemtableMaxRangeDeletions(int p1): ColumnFamilyOptionsInterface</code> has been added, providing control over the maximum number of range deletions in the memtable.
+        <code>org.rocksdb.Options.memtableMaxRangeDeletions(): int</code> has been added, allowing retrieval of the configured maximum number of range deletions in the memtable.

I would only add statements about stuff that was removed because that breaks the users' applications.



In docs/streams/upgrade-guide.html:
> @@ -173,6 +171,21 @@ <h3><a id="streams_api_changes_400" href="#streams_api_changes_400">Streams API
       See <a href="https://cwiki.apache.org/confluence/x/TZCMEw">KIP-1112</a> for more details.
     </p>
 
+    <h3><a id="streams_api_changes_400" href="#streams_api_changes_400">Streams API changes in 4.0.0</a></h3>
+
+    <p>
+        Upgraded RocksDB dependency to version 9.7.3 (from 7.9.2). This upgrade incorporates various improvements and optimizations within RocksDB. However, it also introduces some API changes.
+        The <code>org.rocksdb.AccessHint</code> class, along with its associated methods, has been removed.
+
+        <code>org.rocksdb.Options.setMemtableMaxRangeDeletions(int p1): ColumnFamilyOptionsInterface</code> has been added, providing control over the maximum number of range deletions in the memtable.
+        <code>org.rocksdb.Options.memtableMaxRangeDeletions(): int</code> has been added, allowing retrieval of the configured maximum number of range deletions in the memtable.
+
+        The <code>org.rocksdb.Options.setLogger()</code> method now accepts a <code>LoggerInterface</code> as a parameter instead of the previous <code>Logger</code>.
+        The <code>NO_FILE_CLOSES</code> field has been removed from the <code>org.rocksdb.TickerTypeenum</code>.
+        Some data types used in RocksDB's Java API have been modified. These changes, along with the removed class, field, and new methods, are primarily relevant to users implementing custom RocksDB configurations.
+        These changes are expected to be largely transparent to most Kafka Streams users. However, those employing advanced RocksDB customizations within their Streams applications, particularly through the <code>rocksdb.config.setter</code>, are advised to consult the detailed RocksDB 9.7.3 changelog to ensure a smooth transition and adapt their configurations as needed. Specifically, users leveraging the removed <code>AccessHintclass</code>, the <code>NO_FILE_CLOSES</code> field from <code>TickerType</code>, or relying on the previous signature of <code>setLogger()</code> will need to update their implementations.

The compatibility report also lists the removal of methods of BlockBasedTableConfig which might be important here.



On streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java:
There are a couple of deprecated methods in this file that can be completely removed since 4.0 is a major release. For example baseBackgroundCompactions().



In docs/streams/upgrade-guide.html:
> @@ -139,8 +139,6 @@ <h3 class="anchor-heading"><a id="streams_notable_changes" class="anchor-link"><
         More details about the new config <code>StreamsConfig#TOPOLOGY_OPTIMIZATION_CONFIG</code> can be found in <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-295%3A+Add+Streams+Configuration+Allowing+for+Optional+Topology+Optimization">KIP-295</a>.
     </p>
 
-    <h3><a id="streams_api_changes_400" href="#streams_api_changes_400">Streams API changes in 4.0.0</a></h3>

Still not clear why this line was moved.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

cadonna commented:
@swikarpat 
I would first upgrade to 9.9.3. Experience shows that we should not upgrade to the newest version. In any case, you are welcome to generate a compatibility report and see if they are any breaking changes.

mjsax commented:
Code freeze for the `4.0.0` release happened already. So I don't think we should do anything for `4.0` release. Also agree to Bruno's point about not using a `.0` version.

However, we should evaluate to upgrade again for 4.1.0 release. There is already a ticket for it as placeholder: https://issues.apache.org/jira/browse/KAFKA-18462

ableegoldman commented:
+1 to not using a `.0` version. And while we can't/shouldn't do anything further for 4.0, the branch has already been cut, so we can start thinking about upgrading for 4.1 -- which imo we should start working on ASAP to flush out any weird issues that arise from upgrading to 9.9.3
